cmd/libsnap-confine-private/cgroup-support.c

```diff
@@ -97,3 +103,141 @@ bool sc_cgroup_is_v2() {
+        }
+        attr_path = tmp + 1;
+
+        tmp = strchr(attr_path, ':');
+        if (tmp != NULL) {
+            die("cannot parse cgroup, expected end of line: %s", line);
+        }
+        attr_ctrl_list[-1] = '\0';
+        attr_path[-1] = '\0';
+
+        debug("cgroup presence: id:%s, controllers:%s, path:%s", attr_id, attr_ctrl_list, attr_path);
+
+        /* Find preferred tracking cgroup, either v1 name=systemd or v2. */
+        if (strcmp(attr_id, "0") == 0) {
+            /* We found it but is it mounted for us to see?
+             *
+             * Note that in reality we ought to parse /proc/self/mountinfo and
+             * look but in practice it's either only /sys/fs/cgroup in pure v2
+             * mode or /sys/fs/cgroup/unified in hybrid mode (available since
+             * kernel 4.5). */
+            struct statfs stat_buf;
+            if (statfs(cgroup_dir, &stat_buf) < 0) {
+                if (errno != ENOENT) {
+                    die("cannot statfs %s", cgroup_dir);
+                }
+            } else {
+                if (stat_buf.f_type == CGROUP2_SUPER_MAGIC) {
+                    sc_must_snprintf(path, path_size, "%s%s", cgroup_dir, attr_path);
+                    found = true;
+                    debug("cgroup pure v2 path is %s (preferred)", path);
+                    break;
+                }
+            }
+
+            if (statfs(cgroup_unified_dir, &stat_buf) < 0) {
+                if (errno != ENOENT) {
+                    die("cannot statfs %s", cgroup_unified_dir);
+                }
+            } else {
+                if (stat_buf.f_type == CGROUP2_SUPER_MAGIC) {
+                    sc_must_snprintf(path, path_size, "%s%s", cgroup_unified_dir, attr_path);
+                    debug("cgroup hybrid v2 path is %s (preferred)", path);
+                    found = true;
+                    break;
+                }
+            }
+
+            /* v2 is preferred, don't look more if we found it. */
+            break;
+        } else if (strcmp(attr_ctrl_list, "name=systemd") == 0) {
+            /* We found it but is it mounted for us to see? */
+            struct statfs stat_buf;
+            if (statfs(cgroup_systemd_dir, &stat_buf) < 0) {
+                if (errno != ENOENT) {
+                    die("cannot statfs %s", cgroup_systemd_dir);
+                }
+            } else {
+                if (stat_buf.f_type == CGROUP_SUPER_MAGIC) {
+                    sc_must_snprintf(path, path_size, "%s%s", cgroup_systemd_dir, attr_path);
+                    debug("cgroup v1 name=systemd is %s", path);
+                    found = true;
+                    /* keep looking, maybe there is more */
+                }
+            }
+        }
+    }
+    free(line);
+    fclose(stream);
+    if (!found) {
+        die("cannot find tracking cgroup path");
+    }
+}
+
+void sc_join_sub_cgroup(const char *security_tag, pid_t pid) {
```

Comment:
This could definitely use an overarching comment about why you are creating current_hierarchy_path/security_tag and putting the pid in it.